Conformance testing: Image configurations differ as committed in Docker format

**Description**

Several conformance tests fail under Debian SID with Docker 23.0.1 - the version packaged for Debian 11 (bullseye).  Failing tests (using either VFS or overlay storage):

* `TestConformance/multistage-builtin-args`
* `TestConformance/workdir-owner`
* `TestConformance/env-precedence`

**Steps to reproduce the issue:**
1. On a Debian SID VM with buildah installed from source
2. Install Docker-CE 23.0.1-1~debian.12~bookworm
3. run `make test-conformance`


**Describe the results you received:**
([typical example log](https://cirrus-ci.com/task/5611343332507648?logs=conformance_test#L413))

```
[+1306s]     conformance_test.go:470: 
[+1306s]         	Error Trace:	/var/tmp/go/src/github.com/containers/buildah/tests/conformance/conformance_test.go:470
[+1306s]         	            				/var/tmp/go/src/github.com/containers/buildah/tests/conformance/conformance_test.go:331
[+1306s]         	            				/var/tmp/go/src/github.com/containers/buildah/tests/conformance/conformance_test.go:149
[+1306s]         	Error:      	Image configurations differ as committed in Docker format
[+1306s]         	Test:       	TestConformance/workdir-owner
[+1306s]         	Messages:   	Fields missing from buildah version: moby.buildkit.buildinfo.v1
[+1306s]         	            	Fields which only exist in buildah version: config:AttachStderr config:Entrypoint config:Volumes config:StdinOnce config:Labels config:AttachStdout config:Tty config:OpenStdin config:Domainname config:AttachStdin container_config parent
[+1306s]     conformance_test.go:365: Dockerfile contents:
[+1306s]         FROM alpine
[+1306s]         USER daemon
[+1306s]         WORKDIR /created/directory
[+1306s]         RUN ls /created
[+1306s]         (no final end-of-line)
[+1306s] === RUN   TestConformance/env-precedence
[+1312s]     conformance_test.go:470: 
[+1312s]         	Error Trace:	/var/tmp/go/src/github.com/containers/buildah/tests/conformance/conformance_test.go:470
[+1312s]         	            				/var/tmp/go/src/github.com/containers/buildah/tests/conformance/conformance_test.go:331
[+1312s]         	            				/var/tmp/go/src/github.com/containers/buildah/tests/conformance/conformance_test.go:149
[+1312s]         	Error:      	Image configurations differ as committed in Docker format
[+1312s]         	Test:       	TestConformance/env-precedence
[+1312s]         	Messages:   	Fields missing from buildah version: moby.buildkit.buildinfo.v1
[+1312s]         	            	Fields which only exist in buildah version: config:Labels config:AttachStdin config:AttachStderr config:Tty config:Volumes config:Domainname config:User config:OpenStdin config:StdinOnce config:Entrypoint config:AttachStdout config:WorkingDir container_config parent
[+1312s]     conformance_test.go:353: Context "/var/tmp/go/src/github.com/containers/buildah/tests/conformance/testdata/env/precedence"
[+1312s]     conformance_test.go:365: Dockerfile contents:
[+1312s]         FROM busybox
[+1312s]         ENV a=b
[+1312s]         ENV c=d
[+1312s]         RUN echo E=$E G=$G
[+1312s]         ENV E=E G=G
[+1312s]         RUN echo E=$E G=$G
[+1313s] --- FAIL: TestConformance (1051.79s)
```

**Describe the results you expected:**

Conformance tests should all pass

**Output of `rpm -q buildah` or `apt list buildah`:**

Compiled from source

**Output of `buildah version`:**

```
Version:         1.30.0-dev
Go Version:      go1.19.6
Image Spec:      1.0.2-dev
Runtime Spec:    1.1.0-rc.1
CNI Spec:        1.0.0
libcni Version:  v1.1.2
image Version:   5.24.3-dev
Git Commit:      3f215f9b02a084aefa7e30538bc605a69e91138f
Built:           Thu Mar  2 15:47:55 2023
OS/Arch:         linux/amd64
BuildPlatform:   linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

N/A

**Output of `cat /etc/*release`:**

```
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
VERSION_ID="12.03"
```

**Output of `uname -a`:**

```
Linux cevich-debian-c20230221t162829z-f37f36d12 6.1.0-5-cloud-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.12-1 (2023-02-15) x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
<file doesn't exist>
```

--- a/tests/conformance/README.md
+++ b/tests/conformance/README.md
@@ -15,14 +15,22 @@ Conformance tests use Docker CE to build images to be compared with images built
 
 ## Run conformance tests
 
-You can run all of the tests with go test:
+First, pull base images used by various conformance tests:
 ```
-go test -v -tags "$(./btrfs_tag.sh) $(./btrfs_installed_tag.sh) $(./libdm_tag.sh)" ./tests/conformance
+bash
+docker pull alpine
+docker pull busybox
+docker pull registry.centos.org/centos/centos:centos7
+```
+
+Then you can run all of the tests with go test:
+```
+go test -v -timeout=30m -tags "$(./btrfs_tag.sh) $(./btrfs_installed_tag.sh) $(./libdm_tag.sh)" ./tests/conformance
 ```
 
 If you want to run one of the test cases you can use the "-run" flag:
 ```
-go test -v -tags "$(./btrfs_tag.sh) $(./btrfs_installed_tag.sh) $(./libdm_tag.sh)" -run TestConformance/shell ./tests/conformance
+go test -v -timeout=30m -tags "$(./btrfs_tag.sh) $(./btrfs_installed_tag.sh) $(./libdm_tag.sh)" -run TestConformance/shell ./tests/conformance
 ```
 
 If you also want to build and compare on a line-by-line basis, run:

--- a/tests/conformance/conformance_test.go
+++ b/tests/conformance/conformance_test.go
@@ -26,6 +26,7 @@ import (
 	"github.com/containers/buildah/imagebuildah"
 	"github.com/containers/image/v5/docker/daemon"
 	"github.com/containers/image/v5/image"
+	"github.com/containers/image/v5/manifest"
 	"github.com/containers/image/v5/pkg/compression"
 	istorage "github.com/containers/image/v5/storage"
 	"github.com/containers/image/v5/transports"
@@ -64,6 +65,7 @@ var (
 		"container_config:image",
 		"history",
 		"rootfs:diff_ids",
+		"moby.buildkit.buildinfo.v1",
 	}
 	ociSkip = []string{
 		"created",
@@ -436,7 +438,7 @@ func testConformanceInternalBuild(ctx context.Context, t *testing.T, cwd string,
 	}
 
 	// the report on the buildah image should always be there
-	originalBuildahConfig, ociBuildahConfig, fsBuildah := readReport(t, filepath.Join(buildahDir, t.Name()))
+	_, originalBuildahConfig, ociBuildahConfig, fsBuildah := readReport(t, filepath.Join(buildahDir, t.Name()))
 	if t.Failed() {
 		t.FailNow()
 	}
@@ -460,7 +462,9 @@ func testConformanceInternalBuild(ctx context.Context, t *testing.T, cwd string,
 
 	// the report on the docker image should be there if we expected the build to succeed
 	if !test.withoutDocker {
-		originalDockerConfig, ociDockerConfig, fsDocker = readReport(t, filepath.Join(dockerDir, t.Name()))
+		var mediaType string
+		mediaType, originalDockerConfig, ociDockerConfig, fsDocker = readReport(t, filepath.Join(dockerDir, t.Name()))
+		assert.Equal(t, manifest.DockerV2Schema2MediaType, mediaType, "Image built by docker build didn't use Docker MIME type - tests require update")
 		if t.Failed() {
 			t.FailNow()
 		}
@@ -481,12 +485,12 @@ func testConformanceInternalBuild(ctx context.Context, t *testing.T, cwd string,
 
 	// the report on the imagebuilder image should be there if we expected the build to succeed
 	if compareImagebuilder && !test.withoutImagebuilder {
-		originalDockerConfig, ociDockerConfig, fsDocker = readReport(t, filepath.Join(dockerDir, t.Name()))
+		_, originalDockerConfig, ociDockerConfig, fsDocker = readReport(t, filepath.Join(dockerDir, t.Name()))
 		if t.Failed() {
 			t.FailNow()
 		}
 
-		originalImagebuilderConfig, ociImagebuilderConfig, fsImagebuilder := readReport(t, filepath.Join(imagebuilderDir, t.Name()))
+		_, originalImagebuilderConfig, ociImagebuilderConfig, fsImagebuilder := readReport(t, filepath.Join(imagebuilderDir, t.Name()))
 		if t.Failed() {
 			t.FailNow()
 		}
@@ -819,6 +823,9 @@ func saveReport(ctx context.Context, t *testing.T, ref types.ImageReference, dir
 	img, err := image.FromSource(ctx, nil, src)
 	require.Nil(t, err, "error opening image %q to read its configuration", imageName)
 	closer = img
+	// read the manifest in its original form
+	rawManifest, _, err := src.GetManifest(ctx, nil)
+	require.Nil(t, err, "error reading raw manifest from image %q", imageName)
 	// read the config blob in its original form
 	rawConfig, err := img.ConfigBlob(ctx)
 	require.Nil(t, err, "error reading configuration from image %q", imageName)
@@ -827,8 +834,11 @@ func saveReport(ctx context.Context, t *testing.T, ref types.ImageReference, dir
 	require.Nil(t, err, "error reading OCI-format configuration from image %q", imageName)
 	encodedConfig, err := json.Marshal(ociConfig)
 	require.Nil(t, err, "error encoding OCI-format configuration from image %q for saving", imageName)
+	// save the manifest in its original form
+	err = os.WriteFile(filepath.Join(directory, "manifest.json"), rawManifest, 0644)
+	require.Nil(t, err, "error saving original manifest from image %q", imageName)
 	// save the config blob in the OCI format
-	err = os.WriteFile(filepath.Join(directory, "oci.json"), encodedConfig, 0644)
+	err = os.WriteFile(filepath.Join(directory, "oci-config.json"), encodedConfig, 0644)
 	require.Nil(t, err, "error saving OCI-format configuration from image %q", imageName)
 	// save the config blob in its original format
 	err = os.WriteFile(filepath.Join(directory, "config.json"), rawConfig, 0644)
@@ -1005,7 +1015,17 @@ func applyLayerToFSTree(t *testing.T, layer *Layer, root *FSEntry) {
 }
 
 // read information about the specified image from the specified directory
-func readReport(t *testing.T, directory string) (original, oci, fs map[string]interface{}) {
+func readReport(t *testing.T, directory string) (manifestType string, original, oci, fs map[string]interface{}) {
+	// read the manifest in the as-committed format, whatever that is
+	originalManifest, err := os.ReadFile(filepath.Join(directory, "manifest.json"))
+	require.Nil(t, err, "error reading manifest %q", filepath.Join(directory, "manifest.json"))
+	// dump it into a map
+	manifest := make(map[string]interface{})
+	err = json.Unmarshal(originalManifest, &manifest)
+	require.Nil(t, err, "error decoding manifest %q", filepath.Join(directory, "manifest.json"))
+	if str, ok := manifest["mediaType"].(string); ok {
+		manifestType = str
+	}
 	// read the config in the as-committed (docker) format
 	originalConfig, err := os.ReadFile(filepath.Join(directory, "config.json"))
 	require.Nil(t, err, "error reading configuration file %q", filepath.Join(directory, "config.json"))
@@ -1014,8 +1034,8 @@ func readReport(t *testing.T, directory string) (original, oci, fs map[string]in
 	err = json.Unmarshal(originalConfig, &original)
 	require.Nil(t, err, "error decoding configuration from file %q", filepath.Join(directory, "config.json"))
 	// read the config in converted-to-OCI format
-	ociConfig, err := os.ReadFile(filepath.Join(directory, "oci.json"))
-	require.Nil(t, err, "error reading OCI configuration file %q", filepath.Join(directory, "oci.json"))
+	ociConfig, err := os.ReadFile(filepath.Join(directory, "oci-config.json"))
+	require.Nil(t, err, "error reading OCI configuration file %q", filepath.Join(directory, "oci-config.json"))
 	// dump it into a map
 	oci = make(map[string]interface{})
 	err = json.Unmarshal(ociConfig, &oci)
@@ -1028,7 +1048,7 @@ func readReport(t *testing.T, directory string) (original, oci, fs map[string]in
 	err = json.Unmarshal(fsInfo, &fs)
 	require.Nil(t, err, "error decoding filesystem summary from file %q", filepath.Join(directory, "fs.json"))
 	// return both
-	return original, oci, fs
+	return manifestType, original, oci, fs
 }
 
 // contains is used to check if item is exist in []string or not
@@ -2955,12 +2975,11 @@ var internalTestCases = []testCase{
 		fsSkip:     []string{"(dir):newdir:mtime", "(dir):newdir:(dir):subdir:mtime"},
 	},
 
-	//FIXME: #4639 Fails conformance testing w/ Docker 23.0.1
-	//{
-	//	name:              "multistage-builtin-args",
-	//	dockerfile:        "Dockerfile.margs",
-	//	dockerUseBuildKit: true,
-	//},
+	{
+		name:              "multistage-builtin-args",
+		dockerfile:        "Dockerfile.margs",
+		dockerUseBuildKit: true,
+	},
 
 	{
 		name:       "replace-symlink-with-directory",
@@ -2987,23 +3006,21 @@ var internalTestCases = []testCase{
 		fsSkip:     []string{"(dir):tree:mtime"},
 	},
 
-	//FIXME: #4639 Fails conformance testing w/ Docker 23.0.1
-	//{
-	//	name: "workdir-owner", // from issue #3620
-	//	dockerfileContents: strings.Join([]string{
-	//		`FROM alpine`,
-	//		`USER daemon`,
-	//		`WORKDIR /created/directory`,
-	//		`RUN ls /created`,
-	//	}, "\n"),
-	//	fsSkip:            []string{"(dir):created:mtime", "(dir):created:(dir):directory:mtime"},
-	//	dockerUseBuildKit: true,
-	//},
+	{
+		name: "workdir-owner", // from issue #3620
+		dockerfileContents: strings.Join([]string{
+			`FROM alpine`,
+			`USER daemon`,
+			`WORKDIR /created/directory`,
+			`RUN ls /created`,
+		}, "\n"),
+		fsSkip:            []string{"(dir):created:mtime", "(dir):created:(dir):directory:mtime"},
+		dockerUseBuildKit: true,
+	},
 
-	//FIXME: #4639 Fails conformance testing w/ Docker 23.0.1
-	//{
-	//	name:              "env-precedence",
-	//	contextDir:        "env/precedence",
-	//	dockerUseBuildKit: true,
-	//},
+	{
+		name:              "env-precedence",
+		contextDir:        "env/precedence",
+		dockerUseBuildKit: true,
+	},
 }

